sdks/java/io/amazon-web-services/src/test/java/org/apache/beam/sdk/io/aws/sqs/SqsIOTest.java

```diff
@@ -0,0 +1,138 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.aws.sqs;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.BasicAWSCredentials;
+import com.amazonaws.client.builder.AwsClientBuilder;
+import com.amazonaws.services.sqs.AmazonSQS;
+import com.amazonaws.services.sqs.AmazonSQSClientBuilder;
+import com.amazonaws.services.sqs.model.CreateQueueResult;
+import com.amazonaws.services.sqs.model.Message;
+import com.amazonaws.services.sqs.model.ReceiveMessageResult;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.transforms.Count;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.values.PCollection;
+import org.elasticmq.rest.sqs.SQSRestServer;
+import org.elasticmq.rest.sqs.SQSRestServerBuilder;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExternalResource;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Tests on {@link SqsIO}. */
+@RunWith(JUnit4.class)
+public class SqsIOTest {
+
+  @Rule public TestPipeline pipeline = TestPipeline.create();
+
+  @Rule public EmbeddedSqsServer embeddedSqsRestServer = new EmbeddedSqsServer();
+
+  @Test
+  public void testRead() {
+    final AmazonSQS client = embeddedSqsRestServer.getClient();
+    final String queueUrl = embeddedSqsRestServer.getQueueUrl();
+
+    final PCollection<Message> output =
+        pipeline.apply(SqsIO.read().withQueueUrl(queueUrl).withMaxNumRecords(100));
+
+    PAssert.thatSingleton(output.apply(Count.globally())).isEqualTo(100L);
+
+    for (int i = 0; i < 100; i++) {
+      client.sendMessage(queueUrl, "This is a test");
+    }
+    pipeline.run();
+  }
+
+  @Test
+  public void testWrite() {
+    final AmazonSQS client = embeddedSqsRestServer.getClient();
+    final String queueUrl = embeddedSqsRestServer.getQueueUrl();
+
+    List<String> messages = new ArrayList<>();
+    for (int i = 0; i < 100; i++) {
+      messages.add("This is a test " + i);
+    }
+    pipeline.apply(Create.of(messages)).apply(SqsIO.write().withQueueUrl(queueUrl));
+    pipeline.run().waitUntilFinish();
+
+    List<String> received = new ArrayList<>();
+    while (received.size() < 100) {
+      final ReceiveMessageResult receiveMessageResult = client.receiveMessage(queueUrl);
+
+      if (receiveMessageResult.getMessages() != null) {
+        for (Message message : receiveMessageResult.getMessages()) {
+          received.add(message.getBody());
+        }
+      }
+    }
+    assertEquals(100, received.size());
+    for (String s : received) {
+      assertTrue(s.startsWith("This is a test "));
+    }
```

Comment:
Minor: perhaps slightly more robust might be the following, to ensure we're not counting a redelivery

```
    for (int i = 0; i < 100; i++) {
      assertTrue(received.contains("This is a test " + i));
    }
```

Reply:
no guarantee of ordering with SQS, but the spirit of your comment still applies, will do